### PR TITLE
DRILL-7615: UNION ALL query returns the wrong result for the decimal value

### PR DESCRIPTION
**Pull request by vvysotskyi**
# [DRILL-7615](https://issues.apache.org/jira/browse/DRILL-7615): UNION ALL query returns the wrong result for the decimal value

## Description
Added checks for union operator to create casts for the case when both input columns have a decimal type, but its scale differs. Enhanced calculation of resulting scale and precision for the decimal data type.

## Documentation
NA

## Testing
Added unit tests to check the fix, ran full tests suite.

### CODE REVIEW

File: common/src/main/java/org/apache/drill/common/types/Types.java [hunk 1]
```diff
@@ -131,6 +132,17 @@ public static boolean isIntervalType(MinorType type) {
     }
   }
 
+  /**
+   * Returns true if all specified types are decimal data types.
+   *
+   * @param types type to check
```
Comment (arina-ielchiieva) [inline common/src/main/java/org/apache/drill/common/types/Types.java]:
```suggestion
   * @param types types to check
```

Comment (vvysotskyi) [reply common/src/main/java/org/apache/drill/common/types/Types.java]: Thanks, fixed

File: common/src/main/java/org/apache/drill/common/types/Types.java [hunk 2]
```diff
@@ -131,6 +132,17 @@ public static boolean isIntervalType(MinorType type) {
     }
   }
 
+  /**
+   * Returns true if all specified types are decimal data types.
+   *
+   * @param types types to check
+   * @return true if all specified types are decimal data type.
+   */
+  public static boolean areDecimalTypes(MajorType... types) {
```
Comment (paul-rogers) [inline common/src/main/java/org/apache/drill/common/types/Types.java]: General enough to define as a method? Seems we'd end up having similar methods for many predicates.

File: common/src/main/java/org/apache/drill/common/types/Types.java [hunk 3]
```diff
@@ -795,15 +807,25 @@ public static boolean isSortable(MinorType type) {
...
+        // resulting precision should take into account resulting scale value and be calculated as
+        // sum of two components:
+        // - max integer digits number (precision - scale) for left and right;
+        // - resulting scale.
+        // So for the case of cast(9999 as decimal(4,0)) and cast(1.23 as decimal(3,2))
+        // resulting scale would be Max(0, 2) = 2 and resulting precision would be Max(4 - 0, 3 - 2) + 2 = 6.
+        // In this case, both values would fit into decimal(6, 2): 9999.00, 1.23
+        int precision = Math.max(leftType.getPrecision() - leftType.getScale(), rightType.getPrecision() - rightType.getScale()) + scale;
+        typeBuilder.setPrecision(precision);
+        typeBuilder.setScale(scale);
```
Comment (paul-rogers) [inline common/src/main/java/org/apache/drill/common/types/Types.java]: Nit: please split the long lint.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java [hunk 4]
```diff
@@ -184,38 +184,43 @@ private void createUnionAller(RecordBatch inputBatch) {
...
+      ErrorCollector collector = new ErrorCollectorImpl();
+      // According to input data names, MinorTypes, DataModes, choose to
       // transfer directly,
       // rename columns or
-      // cast data types (Minortype or DataMode)
-      if (container.getSchema().getColumn(index).hasSameTypeAndMode(vvIn.getField())
-          && vvIn.getField().getType().getMinorType() != TypeProtos.MinorType.MAP // Per DRILL-5521, existing bug for map transfer
-          ) {
+      // cast data types (MinorType or DataMode)
+      if (outputField.hasSameTypeAndMode(inField)
```
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java]: Nit: Seems cleaner to add this predicate to `Types`: `isSameTypeAndMode()`, then say:

```
if (Types.isSameTypeAndMode(outputField.getType(), inputField.getType()) {
```

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java [hunk 5]
```diff
@@ -184,38 +184,43 @@ private void createUnionAller(RecordBatch inputBatch) {
...
+          && inField.getType().getMinorType() != TypeProtos.MinorType.MAP // Per DRILL-5521, existing bug for map transfer
+          && (!Types.areDecimalTypes(inField.getType(), outputField.getType())
+              || (Types.areDecimalTypes(outputField.getType(), inField.getType()) // scale should match for decimal data types
+                  && outputField.getType().getScale() == inField.getType().getScale()))) {
```
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java]: If checking precision and scale (hard to follow), then `Types.isSameType()` might be what you want.

File: exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java [hunk 6]
```diff
@@ -184,38 +184,43 @@ private void createUnionAller(RecordBatch inputBatch) {
...
         }
 
-        // If two inputs' MinorTypes are different,
-        // Insert a cast before the Union operation
-        if(inField.getType().getMinorType() != outputField.getType().getMinorType()) {
+        // If two inputs' MinorTypes are different or types are decimal with different scales,
+        // inserts a cast before the Union operation
+        if (inField.getType().getMinorType() != outputField.getType().getMinorType()
+            || (Types.areDecimalTypes(outputField.getType(), inField.getType())
+                && outputField.getType().getScale() != inField.getType().getScale())) {
```
Comment (paul-rogers) [inline exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/union/UnionAllRecordBatch.java]: Would seem to be a general case. Maybe define a `Types.areAssignableTypes(MajorType type1, MajorType type2)` Would return true only if one type is assignable from the other directly. We probably have the same condition in many places.

File: exec/java-exec/src/test/java/org/apache/drill/exec/store/parquet/TestVarlenDecimal.java [hunk 7]
```diff
@@ -139,8 +145,32 @@ public void testWriteReadCsv() throws Exception {
...
+          "union all \n" +
+          "select d from dfs.tmp.t";
+
+      testBuilder()
+          .sqlQuery(query)
+          .unOrdered()
+          .baselineColumns("d")
+          .baselineValues(new BigDecimal("1000.000"))
+          .baselineValues(new BigDecimal("596.000"))
+          .baselineValues(new BigDecimal("999999999999999.000"))
```
Comment (paul-rogers) [inline exec/java-exec/src/test/java/org/apache/drill/exec/store/parquet/TestVarlenDecimal.java]: This tests *values*. Would it make sense to the use the `RowSet` version so you can also verify *types*, which is the focus of this PR?

Comment (arina-ielchiieva) [reply exec/java-exec/src/test/java/org/apache/drill/exec/store/parquet/TestVarlenDecimal.java]: Not sure why this will be complex.
Here is an example:
```
    RowSet actual = client.queryBuilder().sql(sql).rowSet();

    TupleMetadata expectedSchema = new SchemaBuilder()
        // add your schema here 
        .buildSchema();
    RowSet expected = new RowSetBuilder(client.allocator(), expectedSchema)
       // add your columns here
        .build();
    RowSetUtilities.verify(expected, actual);
```

Comment (vvysotskyi) [reply exec/java-exec/src/test/java/org/apache/drill/exec/store/parquet/TestVarlenDecimal.java]: @arina-ielchiieva, thanks for pointing to the example, initially I tried a similar approach, but union operator returns multiple batches, so in the case of using `client.queryBuilder().sql(sql).rowSet()`, it will fail with (`IllegalStateException`)[https://github.com/apache/drill/blob/09b805aea4dafe50555b23945302cf8f6c491de8/exec/java-exec/src/test/java/org/apache/drill/test/QueryBuilder.java#L359]. I also tried to use `rowSetIterator()`, as it was advised in the error message, but it returns `QueryRowSetIterator` which actually caused complications for me. I tried several approaches, one of them is iterating through it and comparing separate row sets, which corresponds to returned batches, but such a check would look obscure since the first batch was empty.
So I ended up with the simpler check for schema using `testBuilder()`.

Comment (arina-ielchiieva) [reply exec/java-exec/src/test/java/org/apache/drill/exec/store/parquet/TestVarlenDecimal.java]: Thanks for explanation.

**Comment by arina-ielchiieva** [comment]
+1 from my side, @paul-rogers is PR ready to go?
